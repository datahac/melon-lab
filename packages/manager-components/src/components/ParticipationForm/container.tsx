import { withFormik } from 'formik';
import { compose, withHandlers } from 'recompose';
import * as Yup from 'yup';
import { divide, multiply } from '~/utils/functionalBigNumber';
import ParticipationForm from './index';

const calculateInputs = (props, field, value) => {
  const { values } = props;

  const totalValue = field === 'total' ? value : values.total;
  const quantityValue = field === 'quantity' ? value : values.quantity;
  const priceValue = field === 'price' ? value : values.price;

  if (field === 'quantity' || field === 'price') {
    const total = multiply(quantityValue, priceValue);

    if (values.total !== total) {
      props.setFieldValue('total', total);
    }
  } else if (field === 'total' || field === 'price') {
    const quantity = divide(totalValue, priceValue);

    if (values.quantity !== quantity) {
      props.setFieldValue('quantity', quantity);
    }
  }
};

const validation = props => {
  const numberFormat = (0).toFixed(props.decimals);
  const minNumber = numberFormat.slice(0, -1) + '1';

  return Yup.object().shape({
    total: Yup.number()
      .min(minNumber, `Minimum total is ${minNumber}`)
      .required('Total is required.'),
    quantity: Yup.number()
      .min(minNumber, `Minimum quantity is ${minNumber}`)
      .required('Quantity is required.'),
    price: Yup.number()
      .min(minNumber, `Minimum price is ${minNumber}`)
      .required('Price is required.'),
  });
};

const withFormValidation = withFormik({
  mapPropsToValues: props => ({ ...props.initialValues }),
  validationSchema: props => validation(props),
  enableReinitialize: true,
  handleSubmit: (values, form) =>
    form.props.onSubmit && form.props.onSubmit(values),
});

const withFormHandler = compose(
  withHandlers({
    onChange: props => event => {
      props.setFieldValue(event.target.name, event.target.value);
      calculateInputs(props, event.target.name, event.target.value);
    },
  }),
);

export default compose(
  withFormValidation,
  withFormHandler,
)(ParticipationForm);
