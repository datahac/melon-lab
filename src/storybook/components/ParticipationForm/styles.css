.participation-form__input {
  margin: 0 0 var(--space-m) 0;
}

.participation-form__toggles {
  width: 100%;
  display: flex;
  padding: var(--space-m) 0;
}

.participation-form__toggle {
  flex: 1 1 auto;
  overflow: hidden;
}
