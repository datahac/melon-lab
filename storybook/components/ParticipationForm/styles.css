.participation-form__input {
  margin: 0 0 var(--space-m) 0;
}

.participation-form__toggles {
  width: 100%;
  display: flex;
  padding: var(--space-m) 0;
}

.participation-form__toggle {
  flex: 1 1 auto;
  overflow: hidden;
}

.participation-form__spinner {
  margin: 0 auto;
}

.participation-form__current-price {
  display: flex;
  flex-direction: row-reverse;
  margin: 0 var(--space-xs) var(--space-xs) 0;
  color: var(--color-grey);
}
