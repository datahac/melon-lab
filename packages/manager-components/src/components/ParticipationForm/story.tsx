import { storiesOf } from '@storybook/react';
import React from 'react';
import ParticipationForm from './container';

const data = {
  initialValues: {
    price: 1.0,
    type: 'Invest',
  },
  decimals: 4,
  setup: true,
  dataValid: true,
  quoteAsset: 'WETH-T',
  fund: {
    sharePrice: 5,
  },
};

storiesOf('Components|Participation Form', module).add('Default', () => {
  return <ParticipationForm {...data} />;
});
