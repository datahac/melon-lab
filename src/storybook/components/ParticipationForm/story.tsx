import { storiesOf } from '@storybook/react';
import React from 'react';
import ParticipationForm from './';

const data = {
  values: {
    price: 1,
    type: 'Invest',
    quantity: 1,
    total: 1,
  },
  errors: {},
  touched: {},
  decimals: 4,
  setup: true,
  quoteAsset: 'WETH',
};

storiesOf('Components|Participation Form', module).add('Default', () => {
  return <ParticipationForm {...data} />;
});
