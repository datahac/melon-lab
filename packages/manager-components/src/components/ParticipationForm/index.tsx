import React, { Fragment, StatelessComponent } from 'react';
import Button from '~/blocks/Button';
import Form from '~/blocks/Form';
import Input from '~/blocks/Input';
import Notification from '~/blocks/Notification';
import Toggle from '~/blocks/Toggle';

import styles from './styles.css';

interface FormValues {
  price: string;
  quantity: string;
  total: string;
  type: string;
}

export interface ParticipationFormProps {
  dataValid: boolean;
  decimals?: number;
  errors?: any;
  handleBlur?: () => void;
  handleSubmit?: () => void;
  onChange?: () => void;
  quoteAsset: string;
  setup: boolean;
  touched?: any;
  values: FormValues;
}

const ParticipationForm: StatelessComponent<ParticipationFormProps> = ({
  dataValid,
  decimals,
  errors,
  handleBlur,
  handleSubmit,
  onChange,
  quoteAsset,
  setup,
  touched,
  values,
}) => {
  const numberPlaceholder = (0).toFixed(decimals);

  return (
    <div className="participation-form">
      <style jsx>{styles}</style>
      <h3>Participation</h3>
      <Form>
        {!dataValid && (
          <Notification isWarning>
            Invest/Redeem not possible when price feed down
          </Notification>
        )}

        {!setup && (
          <div className="participation-form__toggles">
            <div className="participation-form__toggle">
              <Toggle
                name="type"
                value="Invest"
                text="Invest"
                isChecked={values.type === 'Invest'}
                onChange={onChange}
              />
            </div>
            <div className="participation-form__toggle">
              <Toggle
                name="type"
                value="Slices"
                text="Slices"
                isChecked={values.type === 'Slices'}
                onChange={onChange}
              />
            </div>
          </div>
        )}

        <div className="participation-form__input">
          <Input
            value={values.quantity}
            type="number"
            label="Quantity (Shares)"
            name="quantity"
            insideLabel="true"
            placeholder={numberPlaceholder}
            onChange={onChange}
            onBlur={handleBlur}
            required={true}
            formatNumber={true}
            error={touched.quantity && errors.quantity}
            disabled={!dataValid}
          />
        </div>

        {values.type !== 'Slices' && (
          <Fragment>
            <div className="participation-form__input">
              <Input
                value={values.price}
                type="number"
                label={`Price (${quoteAsset})`}
                name="price"
                insideLabel="true"
                placeholder={numberPlaceholder}
                onChange={onChange}
                onBlur={handleBlur}
                required={true}
                formatNumber={true}
                error={touched.price && errors.price}
                disabled={true}
              />
            </div>
            <div className="participation-form__input">
              <Input
                value={values.total}
                type="number"
                label={`Total (${quoteAsset})`}
                name="total"
                insideLabel="true"
                placeholder={numberPlaceholder}
                onChange={onChange}
                onBlur={handleBlur}
                required={true}
                formatNumber={true}
                error={touched.total && errors.total}
                disabled={!dataValid}
              />
            </div>
          </Fragment>
        )}

        <div className="participation-form__input">
          <Button onClick={handleSubmit} type="submit" disabled={!dataValid}>
            Submit request
          </Button>
        </div>
      </Form>
    </div>
  );
};

export default ParticipationForm;
