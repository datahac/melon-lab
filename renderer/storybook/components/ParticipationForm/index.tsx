import React, { Fragment, StatelessComponent } from 'react';
import * as Tm from '@melonproject/token-math';
import Button from '~/blocks/Button';
import Form from '~/blocks/Form';
import Input from '~/blocks/Input';
import Spinner from '~/blocks/Spinner';
import * as R from 'ramda';
import Notification from '~/blocks/Notification';

import styles from './styles.css';

export interface FormValues {
  price: Tm.PriceInterface;
  quantity: Tm.QuantityInterface;
  total: Tm.QuantityInterface;
  type: string;
}

export interface FormErrors {
  quantity?: string;
  total?: string;
  price?: string;
}

export interface ParticipationFormProps {
  decimals?: number;
  touched?: any;
  errors: FormErrors;
  values: FormValues;
  handleBlur?: () => void;
  handleSubmit?: () => void;
  handleChange?: () => void;
  executeRequest?: () => void;
  cancelRequest?: () => void;
  loading?: boolean;
  address?: string;
  sharePrice?: Tm.PriceInterface;
  isWaiting?: boolean;
  readyToExecute?: boolean;
  isInitialRequest?: boolean;
  isExpired?: boolean;
}

const ParticipationForm: StatelessComponent<ParticipationFormProps> = ({
  decimals = 6,
  errors,
  handleBlur,
  handleSubmit,
  handleChange,
  touched,
  values,
  loading,
  address,
  sharePrice,
  isWaiting,
  readyToExecute,
  executeRequest,
  isInitialRequest,
  isExpired,
  cancelRequest,
}) => {
  const numberPlaceholder = (0).toFixed(decimals);

  return (
    <Fragment>
      <style jsx>{styles}</style>
      {loading ? (
        <div className="participation-form__spinner">
          <Spinner icon size="small" />
        </div>
      ) : (
        <div className="participation-form">
          {isWaiting && !isExpired && (
            <Notification isWarning>
              You already requested an investment for this fund. Please wait!
            </Notification>
          )}

          {readyToExecute && (
            <Fragment>
              <Notification isWarning>
                Execute your investment request!
              </Notification>
              <hr />
              <Button onClick={executeRequest}>Execute Request</Button>
            </Fragment>
          )}

          {isExpired && (
            <Fragment>
              <Notification isWarning>Your request is expired!</Notification>
              <hr />
              <Button onClick={cancelRequest}>Cancel Request</Button>
            </Fragment>
          )}

          {!isWaiting && !readyToExecute && (
            <Fragment>
              <Form onSubmit={handleSubmit}>
                <div className="participation-form__input">
                  <Input
                    value={
                      values.quantity && Tm.toFixed(values.quantity, decimals)
                    }
                    type="number"
                    label="Quantity (Shares)"
                    name="quantity"
                    insideLabel
                    placeholder={numberPlaceholder}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required={true}
                    formatNumber={true}
                    error={touched.quantity && errors.quantity}
                    decimals={decimals}
                  />
                </div>
                <div className="participation-form__current-price">
                  <span>
                    Current share price: {sharePrice && Tm.toFixed(sharePrice)}
                  </span>
                </div>

                <div className="participation-form__input">
                  <Input
                    value={values.price && Tm.toFixed(values.price, decimals)}
                    type="number"
                    label={`Max price (${R.path(
                      ['quote', 'token', 'symbol'],
                      values.price,
                    )})`}
                    name="price"
                    insideLabel
                    placeholder={numberPlaceholder}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required={true}
                    formatNumber={true}
                    error={touched.price && errors.price}
                    decimals={decimals}
                    disabled={isInitialRequest}
                  />
                </div>
                <div className="participation-form__input">
                  <Input
                    value={values.total && Tm.toFixed(values.total, decimals)}
                    type="number"
                    label={`Total (${R.path(
                      ['token', 'symbol'],
                      values.total,
                    )})`}
                    name="total"
                    insideLabel
                    placeholder={numberPlaceholder}
                    onChange={handleChange}
                    onBlur={handleBlur}
                    required={true}
                    formatNumber={true}
                    error={touched.total && errors.total}
                    decimals={decimals}
                  />
                </div>

                <div className="participation-form__input">
                  <Button type="submit">Submit request</Button>
                </div>
              </Form>
            </Fragment>
          )}
        </div>
      )}
    </Fragment>
  );
};

export default ParticipationForm;
