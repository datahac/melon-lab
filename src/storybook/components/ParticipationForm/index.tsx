import React, { Fragment, StatelessComponent } from 'react';
import { toFixed } from '@melonproject/token-math/quantity';
import { toFixed as toFixedPrice } from '@melonproject/token-math/price';
import Button from '~/blocks/Button';
import Form from '~/blocks/Form';
import Input from '~/blocks/Input';
import Toggle from '~/blocks/Toggle';
import * as R from 'ramda';

import styles from './styles.css';

interface FormValues {
  price: string;
  quantity: string;
  total: string;
  type: string;
}

export interface ParticipationFormProps {
  decimals?: number;
  sharePrice?: {
    token: {
      symbol: string;
    };
  };
  setup: boolean;
  touched?: any;
  errors?: any;
  values: FormValues;
  handleBlur?: () => void;
  handleSubmit?: () => void;
  handleChange?: () => void;
}

const ParticipationForm: StatelessComponent<ParticipationFormProps> = ({
  decimals = 4,
  errors,
  handleBlur,
  handleSubmit,
  handleChange,
  setup,
  touched,
  values,
}) => {
  const numberPlaceholder = (0).toFixed(decimals);

  return (
    <div className="participation-form">
      <style jsx>{styles}</style>
      <Form onSubmit={handleSubmit}>
        {!setup && (
          <div className="participation-form__toggles">
            <div className="participation-form__toggle">
              <Toggle
                name="type"
                value="Invest"
                text="Invest"
                isChecked={values.type === 'Invest'}
                onChange={handleChange}
              />
            </div>
            <div className="participation-form__toggle">
              <Toggle
                name="type"
                value="Slices"
                text="Slices"
                isChecked={values.type === 'Slices'}
                onChange={handleChange}
              />
            </div>
          </div>
        )}

        <div className="participation-form__input">
          <Input
            value={values.quantity && toFixed(values.quantity, decimals)}
            type="number"
            label="Quantity (Shares)"
            name="quantity"
            insideLabel="true"
            placeholder={numberPlaceholder}
            onChange={handleChange}
            onBlur={handleBlur}
            required={true}
            formatNumber={true}
            error={touched.quantity && errors.quantity}
            decimals={decimals}
          />
        </div>

        {values.type !== 'Slices' && (
          <Fragment>
            <div className="participation-form__input">
              <Input
                value={values.price && toFixedPrice(values.price, decimals)}
                type="number"
                label={`Price (${R.path(
                  ['quote', 'token', 'symbol'],
                  values.price,
                )})`}
                name="price"
                insideLabel="true"
                placeholder={numberPlaceholder}
                onChange={handleChange}
                onBlur={handleBlur}
                required={true}
                formatNumber={true}
                error={touched.price && errors.price}
                disabled={true}
                decimals={decimals}
              />
            </div>
            <div className="participation-form__input">
              <Input
                value={values.total && toFixed(values.total, decimals)}
                type="number"
                label={`Total (${R.path(['token', 'symbol'], values.total)})`}
                name="total"
                insideLabel="true"
                placeholder={numberPlaceholder}
                onChange={handleChange}
                onBlur={handleBlur}
                required={true}
                formatNumber={true}
                error={touched.total && errors.total}
                decimals={decimals}
              />
            </div>
          </Fragment>
        )}

        <div className="participation-form__input">
          <Button type="submit">Submit request</Button>
        </div>
      </Form>
    </div>
  );
};

export default ParticipationForm;
