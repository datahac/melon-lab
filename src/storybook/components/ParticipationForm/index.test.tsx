import React from 'react';
import ParticipationForm from './';

const data = {
  values: {
    price: 1,
    type: 'Invest',
    quantity: 1,
    total: 1,
  },
  errors: {},
  touched: {},
  decimals: 4,
  setup: true,
  quoteAsset: 'WETH',
};

describe('ParticipationForm', () => {
  const defaultElement = <ParticipationForm {...data} />;
  let tree;

  beforeEach(() => {
    tree = mount(defaultElement);
  });

  it('should render correctly', () => {
    expect(tree).toMatchSnapshot();
  });
});
