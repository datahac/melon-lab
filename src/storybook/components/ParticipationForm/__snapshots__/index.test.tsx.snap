// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipationForm should render correctly 1`] = `undefined`;
